packages/sdk/src/oracle/store/write.ts

```diff
@@ -0,0 +1,107 @@
+// Some things to consider:
+// 1. State can be nested, but it should strictly one direction, each class should ideally only operate on its direct state object.
+// 2. The parent can return new components, but its responsible for initializing the state for the child.
+// 3. You can modify this.state in the component thanks to immer, but you cannot set this.state to a new object, only its properties.
+
+export class User {
+  constructor(private state: Partial<state.User>) {}
+  set(data: Partial<state.User>): void {
+    // note that this is done because we cannot replace this.state = data or immer loses visibility to the change.
+    if (data.chainId) this.chainId(data.chainId);
+    if (data.address) this.address(data.address);
+    if (data.signer) this.signer(data.signer);
+  }
+  chainId(chainId: number): void {
+    this.state.chainId = chainId;
+  }
+  address(address: string): void {
+    this.state.address = address;
+  }
+  signer(signer: ethers.Signer): void {
+    this.state.signer = signer;
+  }
+}
+export class Balances {
+  constructor(private state: Partial<state.Balances>) {}
+  set(address: string, amount: ethers.BigNumber): void {
+    this.state[address] = amount;
+  }
+}
+export class Erc20 {
+  constructor(private state: Partial<state.Erc20>) {}
+  props(data: state.Erc20["props"]): void {
+    this.state.props = data;
+  }
+  balance(account: string, amount: ethers.BigNumber): void {
+    if (!this.state.balances) this.state.balances = {};
+    new Balances(this.state.balances).set(account, amount);
+  }
+  allowance(account: string, spender: string, amount: ethers.BigNumber): void {
+    if (!this.state.allowances) this.state.allowances = {};
+    if (!this.state.allowances[spender]) this.state.allowances[spender] = {};
+    new Balances(this.state.allowances[spender]).set(account, amount);
+  }
+}
+export class OptimisticOracle {
+  constructor(private state: Partial<state.OptimisticOracle>) {}
+  address(address: string): void {
+    this.state.address = address;
+  }
+  request(inputRequest: state.Inputs["request"], request: state.Request): void {
+    const id = [
+      inputRequest.requester,
+      inputRequest.identifier,
+      inputRequest.timestamp,
+      inputRequest.ancillaryData,
+    ].join("!");
+    if (!this.state.requests) this.state.requests = {};
+    this.state.requests[id] = request;
+  }
+  defaultLiveness(defaultLiveness: ethers.BigNumber): void {
+    this.state.defaultLiveness = defaultLiveness;
+  }
+}
+export class Chain {
+  constructor(private state: Partial<state.Chain>) {}
+  erc20s(address: string): Erc20 {
+    if (!this.state?.erc20s) this.state.erc20s = {};
+    if (!this.state.erc20s?.[address]) this.state.erc20s[address] = {};
+    return new Erc20(this.state.erc20s[address]);
+  }
+  optimisticOracle(): OptimisticOracle {
+    if (!this.state?.optimisticOracle) this.state.optimisticOracle = {};
+    return new OptimisticOracle(this.state.optimisticOracle);
+  }
```

Comment:
There's a convention in these classes to use the name of the relevant state as the getter/setter name without "get" or "set". This is fine, but in some classes, these are getters and in others they are setters. Do you think this might be confusing for a user? What happens if you want a getter and a setter for the same value in the same class?

Reply:
the point of this class is that these are all setters, so using the term set to prefix calls would be redundant.  The times where functions return a class, are only returning classes that are other setters. You dont have to know you are getting anything when interacting with the writer.

For example, even though this chain of calls technically 'gets' other classes, the user doesnt need to know that and its more like drilling down into a nested state to finally set something.

`store.write((write) => write.chains(1).erc20s("test").allowance("user1", "oracle", BigNumber.from(1)))` 

if we could do nest into data with `?.` syntax for setting, we could probably forgo these classes

Reply:
This makes more sense when you describe it that way. This isn't meant to be a place where you can access state at all.